### PR TITLE
sys/net/sock_util: don't require interface for link-local addr if there is only a single interface

### PR DESCRIPTION
**Pull request by benpicco**
### Contribution description

When the `sock_tl_str2ep()` function is supplied with a link-local address but no interface, if there is only a single interface use that instead of not setting the interface.

This is to match the behavior of the `ping` command.

### Testing procedure

`tests/nanocoap_cli` is a user of this API.

```
> url get coap://[fe80::90a7:a6ff:fe4b:2e32%5]/hello.txt     
url get coap://[fe80::90a7:a6ff:fe4b:2e32%5]/hello.txt
optpos: 0x8072da1
offset 000: Hello RIOT!
```
still works but now you can also do

```
> url get coap://[fe80::90a7:a6ff:fe4b:2e32]/hello.txt
url get coap://[fe80::90a7:a6ff:fe4b:2e32]/hello.txt
optpos: 0x8072da1
offset 000: Hello RIOT!
```

If there is only a single interface.

### CODE REVIEW

File: sys/net/sock/sock_util.c [hunk 1]
```diff
@@ -188,6 +188,17 @@ int _parse_netif(sock_udp_ep_t *ep_out, char *netifstart)
     return (netifend - netifstart);
 }
 
+int _get_default_netif(sock_udp_ep_t *ep_out)
+{
+    netif_t *netif = netif_iter(NULL);
+    if (netif == NULL || netif_iter(netif) != NULL) {
```
Comment (HendrikVE) [inline sys/net/sock/sock_util.c:194]: Why checking again for `netif_iter(netif) != NULL` ? According to doc `netif_iter` returns `NULL, if there is no interface after @p last`, which would be already true for the first call of `netif_iter`.

File: sys/net/sock/sock_util.c [hunk 2]
```diff
@@ -251,6 +262,10 @@ int sock_tl_str2ep(struct _sock_tl_ep *ep_out, const char *str)
 #ifdef SOCK_HAS_IPV6
     if (inet_pton(AF_INET6, hostbuf, ep_out->addr.ipv6) == 1) {
         ep_out->family = AF_INET6;
+        if (*hostend != '%' &&
+            ipv6_addr_is_link_local((ipv6_addr_t *)ep_out->addr.ipv6)) {
+            return _get_default_netif(ep_out);
```
Comment (HendrikVE) [inline sys/net/sock/sock_util.c:267]: Seems like the unittest never triggers this line, so we need an additional test case.

**Review by HendrikVE — CHANGES_REQUESTED** [review]

File: sys/net/sock/sock_util.c (hunk 1 above)
Comment (benpicco) [reply sys/net/sock/sock_util.c:194]: - `netif_iter(NULL)` gives the first interface
 - `netif_iter(netif_iter(NULL))` gives the second interface
 - and so on

This should only set the interface if there is a first, but not a second interface.

File: sys/net/sock/sock_util.c (hunk 2 above)
Comment (benpicco) [reply sys/net/sock/sock_util.c:267]: I added one.

File: tests/netutils/main.c [hunk 3]
```diff
@@ -232,6 +232,28 @@ static void test_sock_tl_name2ep__name_only(void)
...
+{
+    static const ipv6_addr_t a = { {
+            0xfe, 0x80, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00,
+            0xf8, 0xf9, 0xfa, 0xfb, 0xfc, 0xfd, 0xfe, 0xff
+        }
+    };
+    struct _sock_tl_ep ep;
+
+    /* we want to test default netif assignment with a single interface */
+    netif_unregister(&dummy_netif[0].netif);
```
[the remark this answers is not in the document]
Comment (HendrikVE) [reply tests/netutils/main.c:245]: Mh it's a test, so an assertion looks a bit out of place. Would prefer the `n-1` option.

[the remark this answers is not in the document]
Comment (benpicco) [reply tests/netutils/main.c]: Ok, I changed it to remove all but the last interface.

File: sys/net/sock/sock_util.c [hunk 4]
```diff
@@ -188,6 +188,17 @@ int _parse_netif(sock_udp_ep_t *ep_out, char *netifstart)
     return (netifend - netifstart);
 }
 
+int _get_default_netif(sock_udp_ep_t *ep_out)
```
Comment (benpicco) [inline sys/net/sock/sock_util.c:191]:
```suggestion
static int _get_default_netif(sock_udp_ep_t *ep_out)
```

**Review by kfessel — CHANGES_REQUESTED** [review]
The current behavior isn't to not set the interface but it is set to `0` (as it is done for IPv4) the same is happening for non link-local addresses. (this behavior is currently untested for IPv6, but tested for IPv4 in `unittests/sock_util`)
With this PR non link-local addresses without interface return EINVAL (they did not before).
`0` seems a sane choice to me, if there is just one interface (tell me if I am wrong).

Just to repeat the current behavior is `[fe80::90a7:a6ff:fe4b:2e32]` -> `ep.netif=0, ep.port=0, ep.addr set value`

**Comment by benpicco** [comment]
> With this PR non link-local addresses without interface return EINVAL (they did not before).

huh, that would be a bug - this PR only applies to IPv6 link-local addresses, behavior for all other address types remains unchanged.

**Comment by kfessel** [comment]
> > With this PR non link-local addresses without interface return EINVAL (they did not before).
> 
> huh, that would be a bug - this PR only applies to IPv6 link-local addresses, behavior for all other address types remains unchanged.

oops that wasn't the cause when i tested with `unittests/sock_util` (adding the test for the current untested behavior). The cause was `unittests/sock_util` runs without having any netif this will lead to EINVAL.

**Comment by benpicco** [comment]
superseded by #18065

**Pull request closed without merging**